packages/plugin-data/src/commands/force/data/soql/query.ts

```diff
@@ -0,0 +1,224 @@
+      longDescription: messages.getMessage('resultFormatLongDescription'),
+      options: ['human', 'csv', 'json'],
+      default: 'human',
+    }),
+  };
+
+  /**
+   * Define display function that will produce the desired output based on flag resultformat selection
+   *
+   * @protected
+   */
+  protected static readonly result: SfdxResult = {
+    display(): void {
+      const results = asPlainObject(this.data) as DataSoqlQueryResult;
+      let reporter;
+      switch (results.resultFormat as keyof typeof FormatTypes) {
+        case 'human':
+          reporter = new HumanReporter(results, results.columns, this.ux, results.logger);
+          break;
+        case 'json':
+          reporter = new JsonReporter(results, results.columns, this.ux, results.logger);
+          break;
+        case 'csv':
+          reporter = new CsvReporter(results, results.columns, this.ux, results.logger);
+          break;
+        default:
+          throw new Error(`result format is invalid: ${results.resultFormat}`);
+      }
+      // delegate to selected reporter
+      reporter.display();
+    },
+  };
+
+  // Overrides SfdxCommand.  This is ensured since requiresUsername == true
+  protected org!: Org;
+
+  /**
+   * Command run implementation
+   *
+   * Returns either a DataSoqlQueryResult or a SfdxResult.
+   * When the user is using global '--json' flag an instance of SfdxResult si returned.
+   * This is necessary since '--json' flag reports results in the form of SfdxResult
+   * and bypasses the definition of start result. The goal is to have the output
+   * from '--json' and '--resulformat json' be the same.
+   *
+   * The DataSoqlQueryResult is necessary to communicate user selections to the reporters.
+   * The 'this' object available during display() function does not include user input to
+   * the command, which are necessary for reporter selection.
+   *
+   */
+  public async run(): Promise<DataSoqlQueryResult | SfdxResult> {
+    try {
+      if (this.flags.resultformat !== 'json') this.ux.startSpinner(messages.getMessage('queryRunningMessage'));
+      const query = new SoqlQuery();
+      const queryResult: SoqlQueryResult = await query.runSoqlQuery(
+        this.flags.usetoolingapi ? this.org.getConnection().tooling : this.org.getConnection(),
+        this.flags.query,
+        this.logger
+      );
+      const results = {
+        ...queryResult,
+        resultFormat: this.flags.resultformat,
+        json: this.flags.json,
+        logger: this.logger,
+      };
+      return this.normalizeIfJson(results);
+    } finally {
+      if (this.flags.resultformat !== 'json') this.ux.stopSpinner();
+    }
+  }
+
+  /**
+   * This function maps the DataSoqlQueryResult instance to the the well known 'data' property in
+   * SfdxResult class when the user has selected '--json' flag.
+   *
+   * @param results
+   * @private
+   */
+  private normalizeIfJson(results: DataSoqlQueryResult): DataSoqlQueryResult | SfdxResult {
```

Comment:
This whole thing seems wrong to me. You shouldn't need to do anything with the `data` property. What were you running into?

![image](https://user-images.githubusercontent.com/918434/104247050-e05d6780-541b-11eb-8bb9-8b7450879d4e.png)

Reply:
@amphro , This ties into the additional properties I am passing in the result. There are two ways to ask for json results; `resultformat json` and `--json`. When --json is used the display() call is skipped by command or oclif (not sure which). Given that I have added additional properties to the result, this function "restores" the expected result structure when `--json` is given. Doing this ensures the json displayed using resultformat json is exactly the same as when using --json.

In short, varying the output by user selection is difficult to do using `display()` if the data to help shape the output is not available in the standard `this` object.